File: packages/actions/src/step-dispatch.ts
import {
  riskForToolName,
  riskLevelRequiresApproval,
  safeParseDriverTool,
  type ActionStep,
  type CuaActionResult,
  type DriverTool,
  type GoalLoopObservation,
  type ResolvedIntent,
  type RiskLevel,
  type ToolCallTarget,
} from "@handsoff/contracts";

import { gateToolCall } from "./gate-tool-call";

// Pure, React-free step -> tool / risk / dispatch helpers for the autonomous
// loop. They translate an `ActionStep` (the 6 legacy kinds + the full-surface
// `tool_call`) into the driver tool name, risk-relevant target, and the flat
// (tool, args) the generic `driver.call` passthrough (U1) executes — and gate a
// tick against the U2 per-call approval rule. Kept here in `@handsoff/actions`
// (next to the gate) because they are actions-layer execution logic, not React;
// the controller composes them with the live driver.

type ReadyIntent = Extract<ResolvedIntent, { status: "ready" }>;

// Driver tools that click an element (and so get commit-pattern escalation).
const CLICK_TOOLS: ReadonlySet<string> = new Set(["click", "right_click", "double_click"]);

// The driver tool each ActionStep dispatches as. For the full-surface
// `tool_call` step (U3b) this is the tool the model chose verbatim; the legacy
// 6 kinds map to their driver tool so the rule resolver + tests still gate on
// the same vocabulary. Used to key per-call risk (U2).
export function driverToolForStep(step: ActionStep): DriverTool {
  switch (step.kind) {
    case "tool_call": {
      const parsed = safeParseDriverTool(step.tool);
      // An unknown tool name can't be a DriverTool; gate it as the most
      // dangerous via riskForToolName at the call site. Here we surface the
      // closest safe placeholder for the (string) tool name.
      return parsed.success ? parsed.data : "get_window_state";
    }
    case "click_element":
      return "click";
    case "type_text":
      return "type_text";
    case "set_value":
      return "set_value";
    case "launch_app":
      return "launch_app";
    // inspect_window_state / capture_screenshot are read-only perception.
    default:
      return "get_window_state";
  }
}

// The raw tool name a step calls (string — may be outside DRIVER_TOOLS for a
// hallucinated `tool_call`, which `riskForToolName` then gates as mutating).
export function toolNameForStep(step: ActionStep): string {
  return step.kind === "tool_call" ? step.tool : driverToolForStep(step);
}

// The element index a step targets, from its typed target (legacy kinds) or its
// raw driver args (`element_index`, full-surface tool_call).
export function elementIndexForStep(step: ActionStep): number | undefined {
  if (step.kind === "tool_call") {
    const index = step.args["element_index"];
    return typeof index === "number" ? index : undefined;
  }
  if ("target" in step) return step.target.elementIndex;
  return undefined;
}

// Build the risk-relevant target for a click-ish step from the latest snapshot:
// look the element up by index in the perceived AX elements so `riskForToolCall`
// can escalate a *commit* click (Send/Delete/…) to mutating while leaving plain
// navigation clicks free. Only clicks get a target (keys/scroll/etc. carry their
// own risk); absent element metadata leaves the gate to its safe default.
export function toolCallTargetForStep(
  step: ActionStep,
  observation: GoalLoopObservation | undefined,
): ToolCallTarget | undefined {
  const tool = toolNameForStep(step);
  if (!CLICK_TOOLS.has(tool)) return undefined;
  const index = elementIndexForStep(step);
  if (index === undefined) return undefined;
  const element = observation?.state?.elements.find((candidate) => candidate.index === index);
  if (!element) return undefined;
  return {
    element: {
      ...(element.role !== undefined && { role: element.role }),
      ...(element.label !== undefined && { title: element.label, label: element.label }),
      ...(element.value !== undefined && { value: element.value }),
    },
  };
}

// Map any ActionStep to the (tool, args) the generic driver passthrough
// (`driver.call`, U1) executes. The full-surface `tool_call` passes its args
// straight through (the driver's own flat snake_case shape). The legacy 6 kinds
// are translated to flat args from their ActionTarget's surface pid/windowId so
// the rule-resolver path also flows through the single passthrough executor.
export function driverCallForStep(step: ActionStep): {
  tool: string;
  args: Record<string, unknown>;
} {
  if (step.kind === "tool_call") {
    return { tool: step.tool, args: step.args };
  }
  if (step.kind === "launch_app") {
    return {
      tool: "launch_app",
      args: { app_name: step.appName, ...(step.bundleId ? { bundle_id: step.bundleId } : {}) },
    };
  }
  const surface = step.target.surface;
  const base: Record<string, unknown> = {
    ...(surface.pid !== undefined ? { pid: surface.pid } : {}),
    ...(surface.windowId !== undefined ? { window_id: surface.windowId } : {}),
    ...(step.target.elementIndex !== undefined ? { element_index: step.target.elementIndex } : {}),
  };
  switch (step.kind) {
    case "click_element":
      return { tool: "click", args: base };
    case "type_text":
      return { tool: "type_text", args: { ...base, text: step.text } };
    case "set_value":
      return { tool: "set_value", args: { ...base, value: step.value } };
    // inspect_window_state / capture_screenshot → a read-only window probe.
    default:
      return { tool: "get_window_state", args: base };
  }
}

const RISK_RANK: Record<RiskLevel, number> = {
  read_only: 0,
  reversible: 1,
  mutating: 2,
  destructive_external: 3,
};

export function maxRisk(a: RiskLevel, b: RiskLevel): RiskLevel {
  return RISK_RANK[b] > RISK_RANK[a] ? b : a;
}

// The effective risk of a whole one-action-per-tick plan. Risk is the MAX over:
//   - each step's tool-derived risk (U2 `riskForToolCall`, with click element
//     semantics escalating a commit click — Send/Delete/… — to mutating), and
//   - the plan's declared `risk_level`.
// Taking the max means the gate can ESCALATE but the model can never DOWNGRADE
// below what its own tool risk implies (KD3's anti-bypass rule): a model that
// labels a Send click read_only is still gated, while a model that knows a step
// is mutating keeps it gated even when the element label looks benign. The
// per-step max also gates a tick that mixes a free launch with a commit click.
export function planToolRisk(
  plan: ReadyIntent["action_plan"],
  observation: GoalLoopObservation | undefined,
): RiskLevel {
  return plan.action_plan.reduce<RiskLevel>((max, step) => {
    // riskForToolName (not riskForToolCall) so a hallucinated full-surface tool
    // name is gated as mutating rather than throwing — the safe default.
    const risk = riskForToolName(toolNameForStep(step), toolCallTargetForStep(step, observation));
    return maxRisk(max, risk);
  }, plan.risk_level);
}

// Stamp the gate's effective (possibly escalated) risk onto the ready intent so
// the displayed plan + the approval surface agree with the loop's pause: a
// model-declared reversible click on a commit control becomes a mutating plan
// that visibly requires approval. Immutable — returns a new intent.
export function withEffectiveRisk(intent: ReadyIntent, risk: RiskLevel): ReadyIntent {
  if (risk === intent.risk_level) return intent;
  const requires = riskLevelRequiresApproval(risk);
  return {
    ...intent,
    risk_level: risk,
    requires_approval: requires,
    action_plan: {
      ...intent.action_plan,
      risk_level: risk,
      requires_approval: requires,
    },
  };
}

// Run every step through the U2 per-call gate; return the first blocked result
// if any step needs an approval it doesn't have, else null. The gate is derived
// from the tool + target (driverToolForStep already maps a hallucinated
// full-surface tool to the safe get_window_state placeholder; such a step is
// blocked upstream by the resolver), never the model's claim, so a commit step
// (Send/Delete/…) blocks when unapproved.
export function firstBlockedStep(
  steps: readonly ActionStep[],
  observation: GoalLoopObservation | undefined,
  approved: boolean,
): Extract<CuaActionResult, { status: "blocked" }> | null {
  for (const step of steps) {
    const tool = driverToolForStep(step);
    const target = toolCallTargetForStep(step, observation);
    const gate = gateToolCall({ tool, ...(target ? { target } : {}), approved });
    if (!gate.allowed) return gate.result;
  }
  return null;
}

File: apps/sidecar/DirectorSidecar/DirectorSidecar/ActionDispatch/StepDispatch.swift
//
//  StepDispatch.swift
//  DirectorSidecar
//
//  Port of @handsoff/actions step-dispatch.ts — the pure, UI-free step → tool / risk /
//  dispatch helpers for the autonomous loop (Track B). They translate a `Contracts.ActionStep`
//  (the six legacy kinds + the full-surface `tool_call`) into the driver tool name, the
//  risk-relevant target, and the flat (tool, args) the generic `driver.call` passthrough (U1)
//  executes — and fold a plan's effective risk + find the first step that fails the per-call
//  gate. The loop (PORTING.md § Porting Order 3) composes these with the live driver.
//
//  FAITHFUL-PORT DIVERGENCE (typed tool_call): TS types `tool_call.tool` as a string and
//  re-parses it with `safeParseDriverTool`, so `driverToolForStep` has a "hallucinated name →
//  get_window_state placeholder" branch. In Swift the decoded `ActionStep.toolCall` already
//  carries a validated `Contracts.DriverTool` (an off-surface name fails the decode upstream),
//  so that placeholder branch is unreachable for a decoded step — the safe-default for a raw
//  model string still lives in `Contracts.DriverTool.parse` / `ToolRisk.riskForToolName`,
//  which the loop calls BEFORE a step is constructed. See PORTING.md note for Track B.
//

import Foundation

/// Caseless namespace for the step-dispatch helpers (free functions TS-side).
enum StepDispatch {
    // Driver tools that click an element (and so get commit-pattern escalation). Compared by
    // wire name because `toolNameForStep` yields a String (it may be a raw model tool name).
    private static let clickToolNames: Set<String> = ["click", "right_click", "double_click"]

    // Locally-handled tools the loop executes NATIVELY (U3) — never forwarded to `driver.call`.
    // `write_note` is the compose-and-write surface (NoteWriter → ~/Documents/<title>.md + open);
    // it has no cua-driver counterpart. The loop's dispatch checks `isLocalToolStep` and
    // short-circuits these to their native handler.
    static let localToolNames: Set<String> = [Contracts.DriverTool.writeNote.rawValue]

    /// Whether a step dispatches a locally-handled tool (run natively, not via `driver.call`).
    static func isLocalToolStep(_ step: Contracts.ActionStep) -> Bool {
        localToolNames.contains(toolNameForStep(step))
    }

    /// `driverToolForStep`: the driver tool each step dispatches as. A decoded `tool_call`
    /// carries its validated tool verbatim; the legacy six kinds map to their driver tool so
    /// the rule resolver + gate key on the same vocabulary.
    static func driverToolForStep(_ step: Contracts.ActionStep) -> Contracts.DriverTool {
        switch step {
        case let .toolCall(_, _, tool, _): return tool
        case .clickElement: return .click
        case .typeText: return .typeText
        case .setValue: return .setValue
        case .launchApp: return .launchApp
        // inspect_window_state / capture_screenshot are read-only perception.
        case .inspectWindowState, .captureScreenshot: return .getWindowState
        }
    }

    /// `toolNameForStep`: the raw tool name a step calls (the verbatim tool for a `tool_call`,
    /// the mapped driver tool's wire name for a legacy kind).
    static func toolNameForStep(_ step: Contracts.ActionStep) -> String {
        if case let .toolCall(_, _, tool, _) = step { return tool.rawValue }
        return driverToolForStep(step).rawValue
    }

    /// `elementIndexForStep`: the element index a step targets, from its typed target (legacy
    /// kinds) or its raw driver args (`element_index`, full-surface tool_call).
    static func elementIndexForStep(_ step: Contracts.ActionStep) -> Int? {
        if case let .toolCall(_, _, _, args) = step {
            if case let .number(value) = args["element_index"] { return Int(value) }
            return nil
        }
        return step.actionTarget?.elementIndex
    }

    /// `elementTokenForStep`: the driver's stable per-snapshot element handle a `tool_call` targets
    /// (`element_token`), preferred over `element_index` for addressing. Only the generic tool_call
    /// carries it; the legacy kinds address by index only.
    static func elementTokenForStep(_ step: Contracts.ActionStep) -> String? {
        if case let .toolCall(_, _, _, args) = step, case let .string(token) = args["element_token"] {
            return token.isEmpty ? nil : token
        }
        return nil
    }

    /// Whether the step calls a click tool (the tools that get coordinate-fallback escalation).
    static func isClickStep(_ step: Contracts.ActionStep) -> Bool {
        clickToolNames.contains(toolNameForStep(step))
    }

    /// `clickTargetKey`: a stable identity for the clicked ELEMENT, shared by its AX and coordinate
    /// variants so the escalation memory (#158) tracks one target across both addressing paths. Built
    /// from the surface (pid/window_id) + the element handle (token preferred, else index). Nil for a
    /// non-click step or a click that cites neither a token nor an index (untrackable).
    static func clickTargetKey(_ step: Contracts.ActionStep) -> String? {
        guard isClickStep(step) else { return nil }
        let (_, args) = driverCallForStep(step)
        let pid = args["pid"].map(\.signatureJSON) ?? "?"
        let window = args["window_id"].map(\.signatureJSON) ?? "?"
        if let token = elementTokenForStep(step) { return "\(pid):\(window):tok=\(token)" }
        if let index = elementIndexForStep(step) { return "\(pid):\(window):idx=\(index)" }
        return nil
    }

    /// The observed element a click step targets — matched by `element_token` (preferred) or
    /// `element_index` against the latest snapshot. Nil when the element isn't in the observation.
    static func clickedElement(
        for step: Contracts.ActionStep,
        _ observation: Contracts.GoalLoopObservation?
    ) -> Contracts.CuaElement? {
        guard let elements = observation?.state?.elements else { return nil }
        if let token = elementTokenForStep(step),
           let match = elements.first(where: { $0.token == token }) {
            return match
        }
        if let index = elementIndexForStep(step),
           let match = elements.first(where: { $0.index == index }) {
            return match
        }
        return nil
    }

    /// `windowTargetForStep`: the (pid, window_id) a step addresses, read from its flat driver args —
    /// the handle the coordinate fallback needs to screenshot the target window. Nil when either is
    /// absent (an untargeted call).
    static func windowTargetForStep(_ step: Contracts.ActionStep) -> (pid: Int, windowId: Int)? {
        let (_, args) = driverCallForStep(step)
        guard case let .number(pid)? = args["pid"], case let .number(window)? = args["window_id"]
        else { return nil }
        return (Int(pid), Int(window))
    }

    /// Convert an element's frame CENTER (driver global points) into the window-local SCREENSHOT
    /// PIXELS the `click` tool's CGEvent path consumes (#158). The driver returns element frames in
    /// global points but takes click coordinates in the pixel space of the window's screenshot, whose
    /// scale (px ÷ points) is display/driver dependent — so it is derived from a live screenshot's
    /// size vs the window bounds, not assumed. Verified live: System Settings Battery frame
    /// (732,382,71.5,24) + bounds (718,38,715,875) + screenshot 1281×1568 → (89,638), which navigates.
    static func coordinatePixel(
        frame: Contracts.CuaElementFrame,
        bounds: CuaWindowBounds,
        screenshotWidth: Int,
        screenshotHeight: Int
    ) -> (x: Double, y: Double) {
        let scaleX = bounds.width > 0 ? Double(screenshotWidth) / bounds.width : 1
        let scaleY = bounds.height > 0 ? Double(screenshotHeight) / bounds.height : 1
        return ((frame.centerX - bounds.x) * scaleX, (frame.centerY - bounds.y) * scaleY)
    }

    /// `coordinateClickArgs`: the coordinate (CGEvent) variant of an element-targeted click — the SAME
    /// flat args minus the AX addressing (`element_index`/`element_token`), plus the already-converted
    /// `x`/`y` (window-local screenshot pixels from `coordinatePixel`).
    static func coordinateClickArgs(
        for step: Contracts.ActionStep, x: Double, y: Double
    ) -> [String: Contracts.JSONValue] {
        var args = driverCallForStep(step).args
        args.removeValue(forKey: "element_index")
        args.removeValue(forKey: "element_token")
        args["x"] = .number(x)
        args["y"] = .number(y)
        return args
    }

    /// `toolCallTargetForStep`: the risk-relevant target for a click-ish step, looked up by
    /// index in the latest snapshot's perceived AX elements so `riskForToolCall` can escalate
    /// a *commit* click (Send/Delete/…) to mutating. Only clicks get a target; absent element
    /// metadata leaves the gate to its safe default (gate an unidentifiable click).
    static func toolCallTargetForStep(_ step: Contracts.ActionStep,
                                      _ observation: Contracts.GoalLoopObservation?) -> Contracts.ToolCallTarget? {
        guard clickToolNames.contains(toolNameForStep(step)) else { return nil }
        guard let index = elementIndexForStep(step) else { return nil }
        guard let element = observation?.state?.elements.first(where: { $0.index == index }) else { return nil }
        return Contracts.ToolCallTarget(
            element: .init(role: element.role,
                           title: element.label,
                           label: element.label,
                           value: element.value),
            key: nil, keys: nil, pageAction: nil)
    }

    /// `driverCallForStep`: map any step to the (tool, args) the generic driver passthrough
    /// executes. A `tool_call` passes its flat args straight through (the driver's snake_case
    /// shape); the legacy kinds translate to flat args from their target's surface pid/windowId.
    static func driverCallForStep(_ step: Contracts.ActionStep) -> (tool: String, args: [String: Contracts.JSONValue]) {
        switch step {
        case let .toolCall(_, _, tool, args):
            return (tool.rawValue, args)
        case let .launchApp(_, _, appName, bundleId):
            var args: [String: Contracts.JSONValue] = ["app_name": .string(appName)]
            if let bundleId { args["bundle_id"] = .string(bundleId) }
            return ("launch_app", args)
        case let .clickElement(_, _, target):
            return ("click", baseArgs(target))
        case let .typeText(_, _, target, text):
            var args = baseArgs(target)
            args["text"] = .string(text)
            return ("type_text", args)
        case let .setValue(_, _, target, value):
            var args = baseArgs(target)
            args["value"] = .string(value)
            return ("set_value", args)
        // inspect_window_state / capture_screenshot → a read-only window probe.
        case let .inspectWindowState(_, _, target), let .captureScreenshot(_, _, target):
            return ("get_window_state", baseArgs(target))
        }
    }

    private static func baseArgs(_ target: Contracts.ActionTarget) -> [String: Contracts.JSONValue] {
        var base: [String: Contracts.JSONValue] = [:]
        if let pid = target.surface.pid { base["pid"] = .number(Double(pid)) }
        if let windowId = target.surface.windowId { base["window_id"] = .number(Double(windowId)) }
        if let index = target.elementIndex { base["element_index"] = .number(Double(index)) }
        return base
    }

    /// `maxRisk`: the higher-ranked of two risk levels (delegates to the single policy fold).
    static func maxRisk(_ lhs: RiskLevel, _ rhs: RiskLevel) -> RiskLevel {
        RiskLevel.max(lhs, rhs)
    }

    /// `planToolRisk`: the effective risk of a whole one-action-per-tick plan — the MAX over
    /// each step's tool-derived risk (a commit click escalates to mutating) AND the plan's
    /// declared `risk_level`. The max means the gate can ESCALATE but the model can never
    /// DOWNGRADE below what its own tool risk implies (KD3 anti-bypass). `riskForToolName`
    /// (not `riskForToolCall`) so any raw tool name is gated as mutating rather than throwing.
    static func planToolRisk(_ plan: Contracts.ActionPlan,
                             _ observation: Contracts.GoalLoopObservation?) -> RiskLevel {
        plan.actionPlan.reduce(plan.riskLevel) { acc, step in
            let risk = Contracts.ToolRisk.riskForToolName(
                toolNameForStep(step), target: toolCallTargetForStep(step, observation))
            return RiskLevel.max(acc, risk)
        }
    }

    /// `withEffectiveRisk`: stamp the gate's effective (possibly escalated) risk onto the ready
    /// intent so the displayed plan + the approval surface agree with the loop's pause.
    /// Immutable — returns a new intent (the same value when the risk is unchanged).
    static func withEffectiveRisk(_ intent: Contracts.ResolvedIntent.Ready,
                                  risk: RiskLevel) -> Contracts.ResolvedIntent.Ready {
        if risk == intent.riskLevel { return intent }
        let requires = risk.requiresApproval
        let plan = intent.actionPlan.withRisk(risk, requiresApproval: requires)
        return intent.withRisk(risk, requiresApproval: requires, actionPlan: plan)
    }

    /// `firstBlockedStep`: run every step through the per-call gate; return the first blocked
    /// result if any step needs an approval it doesn't have, else nil. The gate is derived from
    /// the tool + target, never the model's claim, so a commit step blocks when unapproved.
    static func firstBlockedStep(_ steps: [Contracts.ActionStep],
                                 _ observation: Contracts.GoalLoopObservation?,
                                 approved: Bool) -> Contracts.CuaActionResult? {
        for step in steps {
            let tool = driverToolForStep(step)
            let target = toolCallTargetForStep(step, observation)
            if let blocked = ToolCallGate.gate(tool: tool, target: target, approved: approved).blockedResult {
                return blocked
            }
        }
        return nil
    }
}

// MARK: - Construction helpers for the decode-only contract types

// The full-shape contract types (ActionPlan, ResolvedIntent.Ready) are decode-only (custom
// `init(from:)`, no memberwise init) so they faithfully reject malformed JSON. `withEffectiveRisk`
// must rebuild them with a re-derived risk + gate, so these extensions add the memberwise inits
// and immutable copy helpers the dispatch layer needs — kept here (not in the contract files) so
// the contract types stay pure decoders.

extension Contracts.ActionStep {
    /// The surface a legacy kind addresses; nil for `launch_app` (no surface) and `tool_call`
    /// (raw flat args, no typed target).
    var actionTarget: Contracts.ActionTarget? {
        switch self {
        case let .inspectWindowState(_, _, target),
             let .clickElement(_, _, target),
             let .captureScreenshot(_, _, target):
            return target
        case let .typeText(_, _, target, _), let .setValue(_, _, target, _):
            return target
        case .launchApp, .toolCall:
            return nil
        }
    }
}

extension Contracts.ActionPlan {
    init(id: String, summary: String, riskLevel: RiskLevel, requiresApproval: Bool,
         targetAgent: Contracts.TargetAgent, actionPlan: [Contracts.ActionStep]) {
        self.id = id
        self.summary = summary
        self.riskLevel = riskLevel
        self.requiresApproval = requiresApproval
        self.targetAgent = targetAgent
        self.actionPlan = actionPlan
    }

    /// A copy with a re-derived risk + gate (every other field unchanged).
    func withRisk(_ riskLevel: RiskLevel, requiresApproval: Bool) -> Self {
        Self(id: id, summary: summary, riskLevel: riskLevel, requiresApproval: requiresApproval,
             targetAgent: targetAgent, actionPlan: actionPlan)
    }
}

extension Contracts.ResolvedIntent.Ready {
    init(id: String, input: Contracts.IntentInput, intentType: Contracts.IntentType,
         referent: Contracts.SelectedReferent?, constraints: [String], riskLevel: RiskLevel,
         requiresApproval: Bool, targetAgent: Contracts.TargetAgent,
         actionPlan: Contracts.ActionPlan, createdAt: String) {
        self.id = id
        self.input = input
        self.intentType = intentType
        self.referent = referent
        self.constraints = constraints
        self.riskLevel = riskLevel
        self.requiresApproval = requiresApproval
        self.targetAgent = targetAgent
        self.actionPlan = actionPlan
        self.createdAt = createdAt
    }

    /// A copy with a re-derived risk + gate and the matching re-gated plan.
    func withRisk(_ riskLevel: RiskLevel, requiresApproval: Bool,
                  actionPlan: Contracts.ActionPlan) -> Self {
        Self(id: id, input: input, intentType: intentType, referent: referent,
             constraints: constraints, riskLevel: riskLevel, requiresApproval: requiresApproval,
             targetAgent: targetAgent, actionPlan: actionPlan, createdAt: createdAt)
    }
}
